File: ios/BundleManager.swift
import Foundation
import ZIPFoundation

/// Manages bundle storage and switching on iOS.
///
/// Storage location (app-private Documents, excluded from iCloud backup):
///   <Documents>/ota/bundles/<hash>/main.jsbundle
class BundleManager {

    /// Delta patch format this SDK understands. Must match
    /// DELTA_MANIFEST_VERSION in packages/backend/src/services/DeltaService.ts.
    /// A mismatch makes the patch fail cleanly and the caller fall back to a
    /// full download.
    static let deltaManifestVersion = 2

    private let prefs: OTAPrefs

    private var otaRoot: URL {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let root = docs.appendingPathComponent("ota/bundles")
        try? FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        return root
    }

    init(prefs: OTAPrefs) {
        self.prefs = prefs
    }

    /// Destination for a downloaded ZIP before extraction.
    func zipURL(forHash hash: String) -> URL {
        otaRoot.appendingPathComponent("\(hash)/download.zip")
    }

    /// Extract ZIP and store the JS bundle. Returns the bundle file path.
    func storePendingBundle(zipURL: URL, hash: String, bundleId: String) throws -> String {
        let bundleDir = otaRoot.appendingPathComponent(hash)
        try FileManager.default.createDirectory(at: bundleDir, withIntermediateDirectories: true)

        // Unzip
        try FileManager.default.unzipItem(at: zipURL, to: bundleDir)
        try? FileManager.default.removeItem(at: zipURL)

        // Find main JS bundle
        guard let bundleFile = findBundleFile(in: bundleDir) else {
            throw NSError(domain: "OTABundleManager", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "No JS bundle found in ZIP"])
        }

        let bundlePath = bundleFile.path
        prefs.pendingBundlePath = bundlePath
        prefs.pendingBundleHash = hash
        prefs.pendingBundleId   = bundleId

        NSLog("[OTA] Pending bundle stored: %@", bundlePath)
        return bundlePath
    }

    /// Apply a delta patch ZIP on top of an existing bundle directory.
    ///
    /// The patch ZIP (produced by DeltaService on the server) contains:
    /// - `_delta_manifest.json` — version, per-file entries and deletions
    /// - `bin/<n>` — bsdiff patches, applied against the base file
    /// - `raw/<n>` — whole replacement files
    ///
    /// Most changed files arrive as bsdiff patches, so a one-line JS edit in a
    /// multi-megabyte bundle is a few kilobytes on the wire.
    ///
    /// Throws on any inconsistency; the caller discards the partial result and
    /// falls back to downloading the full bundle.
    func applyDeltaPatch(patchZipURL: URL, fromHash: String, toHash: String, bundleId: String) throws -> String {
        let fm = FileManager.default
        let baseBundleDir = otaRoot.appendingPathComponent(fromHash)
        guard fm.fileExists(atPath: baseBundleDir.path) else {
            throw NSError(domain: "OTABundleManager", code: 2,
                          userInfo: [NSLocalizedDescriptionKey: "Base bundle dir not found for hash \(fromHash)"])
        }

        let targetDir = otaRoot.appendingPathComponent(toHash)
        let scratchDir = otaRoot.appendingPathComponent("patch-\(toHash)")

        do {
            try fm.createDirectory(at: targetDir, withIntermediateDirectories: true)
            try fm.createDirectory(at: scratchDir, withIntermediateDirectories: true)

            // 1. Seed target with all files from base
            let baseContents = try fm.contentsOfDirectory(at: baseBundleDir, includingPropertiesForKeys: nil)
            for item in baseContents {
                let dest = targetDir.appendingPathComponent(item.lastPathComponent)
                if fm.fileExists(atPath: dest.path) { try fm.removeItem(at: dest) }
                try fm.copyItem(at: item, to: dest)
            }

            // 2. Unpack the patch somewhere separate — its entry names
            //    (bin/0, raw/1) are not bundle paths and must not land in the
            //    bundle directory.
            try fm.unzipItem(at: patchZipURL, to: scratchDir)
            try? fm.removeItem(at: patchZipURL)

            let manifestURL = scratchDir.appendingPathComponent("_delta_manifest.json")
            guard fm.fileExists(atPath: manifestURL.path),
                  let manifestData = try? Data(contentsOf: manifestURL),
                  let manifest = try JSONSerialization.jsonObject(with: manifestData) as? [String: Any] else {
                throw NSError(domain: "OTABundleManager", code: 4,
                              userInfo: [NSLocalizedDescriptionKey: "Patch is missing or has an unreadable _delta_manifest.json"])
            }

            let version = manifest["version"] as? Int ?? 1
            guard version == BundleManager.deltaManifestVersion else {
                throw NSError(domain: "OTABundleManager", code: 5,
                              userInfo: [NSLocalizedDescriptionKey:
                                "Unsupported delta manifest version \(version) (this SDK understands \(BundleManager.deltaManifestVersion))"])
            }

            // 3. Reconstruct every changed file
            if let files = manifest["files"] as? [[String: Any]] {
                for entry in files {
                    try applyDeltaFile(entry: entry,
                                       baseDir: baseBundleDir,
                                       scratchDir: scratchDir,
                                       targetDir: targetDir)
                }
            }

            // 4. Deletions
            if let deleted = manifest["deleted"] as? [String] {
                for relativePath in deleted {
                    let victim = try safeChild(of: targetDir, relativePath: relativePath)
                    if fm.fileExists(atPath: victim.path) {
                        try? fm.removeItem(at: victim)
                        NSLog("[OTA] Delta: deleted %@", relativePath)
                    }
                }
            }

            // 5. Find JS bundle
            guard let bundleFile = findBundleFile(in: targetDir) else {
                throw NSError(domain: "OTABundleManager", code: 3,
                              userInfo: [NSLocalizedDescriptionKey: "No JS bundle after patching to \(toHash)"])
            }

            try? fm.removeItem(at: scratchDir)

            let bundlePath = bundleFile.path
            prefs.pendingBundlePath = bundlePath
            prefs.pendingBundleHash = toHash
            prefs.pendingBundleId   = bundleId

            NSLog("[OTA] Delta applied: %@ (from=%@ to=%@)", bundlePath, fromHash, toHash)
            return bundlePath
        } catch {
            // Leave nothing half-patched behind for the next attempt to trip on.
            try? fm.removeItem(at: targetDir)
            try? fm.removeItem(at: scratchDir)
            throw error
        }
    }

    /// Reconstruct one file described by a manifest entry.
    private func applyDeltaFile(entry: [String: Any],
                                baseDir: URL,
                                scratchDir: URL,
                                targetDir: URL) throws {
        guard let path = entry["path"] as? String,
              let mode = entry["mode"] as? String,
              let entryName = entry["entry"] as? String,
              let expectedSize = entry["newSize"] as? Int,
              let expectedSha = entry["newSha256"] as? String else {
            throw NSError(domain: "OTABundleManager", code: 6,
                          userInfo: [NSLocalizedDescriptionKey: "Malformed delta manifest entry"])
        }

        let payloadURL = try safeChild(of: scratchDir, relativePath: entryName)
        guard let payload = try? Data(contentsOf: payloadURL) else {
            throw NSError(domain: "OTABundleManager", code: 7,
                          userInfo: [NSLocalizedDescriptionKey: "Patch entry \"\(entryName)\" missing for \(path)"])
        }

        let reconstructed: Data
        switch mode {
        case "bsdiff":
            let baseURL = try safeChild(of: baseDir, relativePath: path)
            guard let baseData = try? Data(contentsOf: baseURL) else {
                throw NSError(domain: "OTABundleManager", code: 8,
                              userInfo: [NSLocalizedDescriptionKey:
                                "Base bundle is missing \"\(path)\", required by a bsdiff entry"])
            }
            reconstructed = try BsPatch.apply(oldData: baseData, patchData: payload)
        case "raw":
            reconstructed = payload
        default:
            throw NSError(domain: "OTABundleManager", code: 9,
                          userInfo: [NSLocalizedDescriptionKey: "Unknown delta mode \"\(mode)\" for \(path)"])
        }

        guard reconstructed.count == expectedSize else {
            throw NSError(domain: "OTABundleManager", code: 10,
                          userInfo: [NSLocalizedDescriptionKey:
                            "Reconstructed \"\(path)\" is \(reconstructed.count) bytes, expected \(expectedSize)"])
        }

        guard HashVerifier.sha256Hex(reconstructed).caseInsensitiveCompare(expectedSha) == .orderedSame else {
            throw NSError(domain: "OTABundleManager", code: 11,
                          userInfo: [NSLocalizedDescriptionKey: "Reconstructed \"\(path)\" failed its hash check"])
        }

        let destination = try safeChild(of: targetDir, relativePath: path)
        try FileManager.default.createDirectory(at: destination.deletingLastPathComponent(),
                                                withIntermediateDirectories: true)
        try reconstructed.write(to: destination)
    }

    /// Resolve a relative path inside `parent`, refusing anything that escapes it.
    /// A manifest is server-supplied data, so "../.." has to be treated as
    /// hostile even though we generated it.
    private func safeChild(of parent: URL, relativePath: String) throws -> URL {
        let child = parent.appendingPathComponent(relativePath).standardizedFileURL
        let root = parent.standardizedFileURL
        guard child.path == root.path || child.path.hasPrefix(root.path + "/") else {
            throw NSError(domain: "OTABundleManager", code: 12,
                          userInfo: [NSLocalizedDescriptionKey:
                            "Path traversal detected in delta manifest: \(relativePath)"])
        }
        return child
    }

    /// Promote pending → active.
    func applyPendingBundle() {
        guard let pendingPath = prefs.pendingBundlePath else { return }
        let pendingId = prefs.pendingBundleId
        prefs.activeBundlePath = pendingPath
        prefs.activeBundleHash = prefs.pendingBundleHash
        // Carry the id across too — clearPending() is about to drop it, and it
        // is the only human-readable handle on what the device is running.
        prefs.activeBundleId   = pendingId
        prefs.clearPending()
        NSLog("[OTA] Bundle applied: %@ (id=%@)", pendingPath, pendingId)
    }

    /// Get active bundle URL for RCTBridge. Returns nil to use embedded asset.
    func getActiveBundleURL() -> URL? {
        guard let path = prefs.activeBundlePath else { return nil }
        let url = URL(fileURLWithPath: path)
        return FileManager.default.fileExists(atPath: path) ? url : nil
    }

    /// Full rollback — clear all OTA state.
    func rollback() {
        NSLog("[OTA] Rolling back to embedded bundle")
        prefs.activeBundlePath = nil
        prefs.activeBundleHash = ""
        prefs.activeBundleId   = ""
        prefs.clearPending()
        prefs.resetCrashCount()
    }

    /// Delete all bundles except the currently active one.
    func cleanOldBundles() {
        let activeHash = prefs.activeBundleHash
        guard let contents = try? FileManager.default.contentsOfDirectory(
            at: otaRoot, includingPropertiesForKeys: nil
        ) else { return }

        for dir in contents where dir.lastPathComponent != activeHash {
            try? FileManager.default.removeItem(at: dir)
            NSLog("[OTA] Cleaned old bundle: %@", dir.lastPathComponent)
        }
    }

    // MARK: - Private

    private func findBundleFile(in dir: URL) -> URL? {
        let names = ["main.jsbundle", "index.ios.bundle", "index.bundle"]
        for name in names {
            let candidate = dir.appendingPathComponent(name)
            if FileManager.default.fileExists(atPath: candidate.path) {
                return candidate
            }
        }
        // Recursive search fallback
        let enumerator = FileManager.default.enumerator(at: dir, includingPropertiesForKeys: nil)
        while let file = enumerator?.nextObject() as? URL {
            if file.pathExtension == "jsbundle" || file.lastPathComponent.hasSuffix(".bundle") {
                return file
            }
        }
        return nil
    }
}

File: android/src/main/java/com/otasdk/BundleManager.kt
package com.otasdk

import android.content.Context
import android.util.Log
import java.io.File

/**
 * Manages bundle storage and switching on Android.
 *
 * Bundle lifecycle:
 *  downloaded → verified → stored as pending → applied  on restart → marked active
 *
 * Storage locations (app-private, no permissions needed):
 *   /data/data/<packageName>/files/ota/bundles/<hash>/index.android.bundle
 */
class BundleManager(private val context: Context, private val prefs: OTAPrefs) {

    companion object {
        private const val TAG = "OTA-BundleManager"
        private const val OTA_DIR = "ota/bundles"

        /**
         * Delta patch format this SDK understands. Must match
         * DELTA_MANIFEST_VERSION in packages/backend/src/services/DeltaService.ts.
         * A mismatch makes the patch fail cleanly and the caller fall back to a
         * full download.
         */
        private const val DELTA_MANIFEST_VERSION = 2
    }

    private val otaRoot: File
        get() = File(context.filesDir, OTA_DIR).also { it.mkdirs() }

    /**
     * Destination file for a bundle identified by its hash.
     */
    fun bundleFileForHash(hash: String): File =
        File(otaRoot, "$hash/index.android.bundle")

    /**
     * Destination file for a downloaded ZIP before extraction.
     */
    fun zipFileForHash(hash: String): File =
        File(otaRoot, "$hash/download.zip")

    /**
     * Store a downloaded + verified ZIP to the OTA bundle directory.
     * Unzips and sets up the bundle path for use by React Native.
     *
     * @return path to the extracted JS bundle file
     */
    fun storePendingBundle(
        zipFile: File,
        hash: String,
        bundleId: String,
    ): String {
        val bundleDir = File(otaRoot, hash)
        bundleDir.mkdirs()

        // Extract ZIP contents into bundleDir
        unzip(zipFile, bundleDir)

        // Find the main JS bundle file (index.android.bundle or main.jsbundle)
        val bundleFile = findBundleFile(bundleDir)
            ?: throw IllegalStateException("No JS bundle found in ZIP for hash $hash")

        Log.d(TAG, "Pending bundle stored: ${bundleFile.absolutePath}")

        // Record pending state
        prefs.pendingBundlePath = bundleFile.absolutePath
        prefs.pendingBundleHash = hash
        prefs.pendingBundleId = bundleId

        return bundleFile.absolutePath
    }

    /**
     * Promote pending bundle to active.
     * Called after a successful restart on the new bundle.
     */
    fun applyPendingBundle() {
        val pendingPath = prefs.pendingBundlePath ?: return
        val pendingHash = prefs.pendingBundleHash

        prefs.activeBundlePath = pendingPath
        prefs.activeBundleHash = pendingHash
        prefs.clearPending()

        Log.d(TAG, "Bundle applied: $pendingPath (hash=$pendingHash)")
    }

    /**
     * Apply a delta patch ZIP on top of an existing bundle directory.
     *
     * The patch ZIP (produced by DeltaService on the server) contains:
     * - `_delta_manifest.json` — version, per-file entries and deletions
     * - `bin/<n>` — bsdiff patches, applied against the base file
     * - `raw/<n>` — whole replacement files
     *
     * Most changed files arrive as bsdiff patches: a one-line JS edit in a
     * multi-megabyte bundle is a few kilobytes on the wire rather than the
     * whole file.
     *
     * Steps:
     *  1. Copy the base bundle directory into a new dir keyed by toHash
     *  2. Extract the patch ZIP to a scratch directory
     *  3. Reconstruct each file listed in the manifest, verifying its hash
     *  4. Delete files the manifest marks as removed
     *  5. Record the patched bundle as pending
     *
     * Throws on any inconsistency. The caller catches, discards the partial
     * result, and falls back to downloading the full bundle.
     *
     * @return Path to the JS bundle file after patching
     */
    fun applyDeltaPatch(
        patchZipFile: File,
        fromHash: String,
        toHash: String,
        bundleId: String,
    ): String {
        val baseBundleDir = File(otaRoot, fromHash)
        if (!baseBundleDir.exists()) {
            throw IllegalStateException(
                "Cannot apply delta: base bundle dir not found for hash $fromHash"
            )
        }

        val targetDir = File(otaRoot, toHash)
        val scratchDir = File(otaRoot, "patch-$toHash")

        try {
            targetDir.mkdirs()
            scratchDir.mkdirs()

            // 1. Seed the target dir from the base
            baseBundleDir.copyRecursively(targetDir, overwrite = true)

            // 2. Unpack the patch somewhere separate — its entry names
            //    (bin/0, raw/1) are not bundle paths and must not land in the
            //    bundle directory.
            unzip(patchZipFile, scratchDir)

            val manifestFile = File(scratchDir, "_delta_manifest.json")
            if (!manifestFile.exists()) {
                throw IllegalStateException("Patch is missing _delta_manifest.json")
            }

            val manifest = org.json.JSONObject(manifestFile.readText())
            val version = manifest.optInt("version", 1)
            if (version != DELTA_MANIFEST_VERSION) {
                throw IllegalStateException(
                    "Unsupported delta manifest version $version " +
                        "(this SDK understands $DELTA_MANIFEST_VERSION)"
                )
            }

            // 3. Reconstruct every changed file
            val files = manifest.optJSONArray("files")
            if (files != null) {
                for (i in 0 until files.length()) {
                    applyDeltaFile(files.getJSONObject(i), baseBundleDir, scratchDir, targetDir)
                }
            }

            // 4. Deletions
            val deleted = manifest.optJSONArray("deleted")
            if (deleted != null) {
                for (i in 0 until deleted.length()) {
                    val relativePath = deleted.getString(i)
                    val victim = safeChild(targetDir, relativePath)
                    if (victim.exists()) {
                        victim.delete()
                        Log.d(TAG, "Delta: deleted $relativePath")
                    }
                }
            }

            // 5. Locate JS bundle
            val bundleFile = findBundleFile(targetDir)
                ?: throw IllegalStateException("No JS bundle found after patching to hash $toHash")

            Log.d(TAG, "Delta applied: ${bundleFile.absolutePath} (from=$fromHash to=$toHash)")

            prefs.pendingBundlePath = bundleFile.absolutePath
            prefs.pendingBundleHash = toHash
            prefs.pendingBundleId   = bundleId

            return bundleFile.absolutePath
        } catch (e: Exception) {
            // Leave nothing half-patched behind for the next attempt to trip on.
            targetDir.deleteRecursively()
            throw e
        } finally {
            scratchDir.deleteRecursively()
        }
    }

    /** Reconstruct one file described by a manifest entry. */
    private fun applyDeltaFile(
        entry: org.json.JSONObject,
        baseDir: File,
        scratchDir: File,
        targetDir: File,
    ) {
        val path = entry.getString("path")
        val mode = entry.getString("mode")
        val entryName = entry.getString("entry")
        val expectedSize = entry.getLong("newSize")
        val expectedSha = entry.getString("newSha256")

        val payloadFile = safeChild(scratchDir, entryName)
        if (!payloadFile.exists()) {
            throw IllegalStateException("Patch entry \"$entryName\" missing for $path")
        }
        val payload = payloadFile.readBytes()

        val reconstructed = when (mode) {
            "bsdiff" -> {
                val baseFile = safeChild(baseDir, path)
                if (!baseFile.exists()) {
                    throw IllegalStateException(
                        "Base bundle is missing \"$path\", required by a bsdiff entry"
                    )
                }
                BsPatch.apply(baseFile.readBytes(), payload)
            }
            "raw" -> payload
            else -> throw IllegalStateException("Unknown delta mode \"$mode\" for $path")
        }

        if (reconstructed.size.toLong() != expectedSize) {
            throw IllegalStateException(
                "Reconstructed \"$path\" is ${reconstructed.size} bytes, expected $expectedSize"
            )
        }

        val actualSha = HashVerifier.sha256Hex(reconstructed)
        if (!actualSha.equals(expectedSha, ignoreCase = true)) {
            throw IllegalStateException("Reconstructed \"$path\" failed its hash check")
        }

        val destination = safeChild(targetDir, path)
        destination.parentFile?.mkdirs()
        destination.writeBytes(reconstructed)
    }

    /**
     * Resolve a relative path inside [parent], refusing anything that escapes it.
     * A manifest is server-supplied data, so "../../etc/passwd" has to be
     * treated as hostile even though we generated it.
     */
    private fun safeChild(parent: File, relativePath: String): File {
        val child = File(parent, relativePath)
        require(child.canonicalPath.startsWith(parent.canonicalPath + File.separator) ||
                child.canonicalPath == parent.canonicalPath) {
            "Path traversal detected in delta manifest: $relativePath"
        }
        return child
    }

    /**
     * Get the active JS bundle path.
     * Returns null if no OTA bundle has ever been applied (use bundled asset).
     */
    fun getActiveBundlePath(): String? = prefs.activeBundlePath

    /**
     * Roll back to the original bundled asset (clear all OTA state).
     */
    fun rollback() {
        Log.w(TAG, "Rolling back — clearing OTA bundle state")
        prefs.activeBundlePath = null
        prefs.activeBundleHash = ""
        prefs.clearPending()
        prefs.resetCrashCount()
    }

    /**
     * Delete all OTA bundles except the current active one.
     * Called periodically to clean up old bundles.
     */
    fun cleanOldBundles() {
        val activeHash = prefs.activeBundleHash
        otaRoot.listFiles()?.forEach { dir ->
            if (dir.isDirectory && dir.name != activeHash) {
                Log.d(TAG, "Cleaning old bundle: ${dir.name}")
                dir.deleteRecursively()
            }
        }
    }

    // ── ZIP extraction ──────────────────────────────────────────────
    private fun unzip(zipFile: File, destDir: File) {
        java.util.zip.ZipInputStream(zipFile.inputStream().buffered()).use { zis ->
            var entry = zis.nextEntry
            while (entry != null) {
                val target = File(destDir, entry.name)
                // Security: prevent path traversal
                require(target.canonicalPath.startsWith(destDir.canonicalPath)) {
                    "ZIP path traversal detected: ${entry.name}"
                }
                if (entry.isDirectory) {
                    target.mkdirs()
                } else {
                    target.parentFile?.mkdirs()
                    target.outputStream().use { out -> zis.copyTo(out) }
                }
                zis.closeEntry()
                entry = zis.nextEntry
            }
        }
        zipFile.delete() // Remove ZIP after extraction
    }

    private fun findBundleFile(dir: File): File? =
        sequenceOf("index.android.bundle", "main.jsbundle", "index.bundle")
            .map { File(dir, it) }
            .firstOrNull { it.exists() }
            ?: dir.walk().firstOrNull { it.name.endsWith(".bundle") }

}
